### ISSUE TITLE
Ambiguous constructor call. The constructors BuildNotification() and BuildNotification() both match.

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-xtend#897]
```
Description	Resource	Path	Location	Type
Ambiguous constructor call.
The constructors
	BuildNotification() and
	BuildNotification()
both match.	TestLangLSPExtension.xtend	/org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide	line: 80 /org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/TestLangLSPExtension.xtend	Xtend Problem

```

**Comment by cdietrich** [comment]
@szarnekow looks like there is something wrong with our fix for the ambiguous constructor calls

**Comment by cdietrich** [comment]
@szarnekow looks like we compare a candidate with itself and think thats ambigous ;)

![Bildschirmfoto vom 2019-07-26 09-00-08](https://user-images.githubusercontent.com/238736/61932511-e5a6c800-af83-11e9-97b3-6c2bd6eb9892.png)

**Comment by cdietrich** [comment]
why 2 descriptions for the same constructor?

![Bildschirmfoto vom 2019-07-26 09-08-39](https://user-images.githubusercontent.com/238736/61932982-002d7100-af85-11e9-8df5-9101c8fa1d45.png)

**Comment by cdietrich** [comment]
`org.eclipse.xtext.xbase.scoping.batch.ConstructorTypeScopeWrapper.getElements(QualifiedName)` finds the type twice. this is caused by
`org.eclipse.xtext.xbase.scoping.batch.NestedTypesScope.getLocalElementsByName(QualifiedName)` which adds the nested type twice

**Comment by cdietrich** [comment]
here is a minimal reproducing sample

```
interface Sample {
	class Demo {
		
	}
	static class SyampleImpl implements Sample {
		def x() {
			new Demo
		}
	}
}
```

**Comment by cdietrich** [comment]
maybe

```
	@Override
	protected Iterable<IEObjectDescription> getLocalElementsByName(QualifiedName name) {
		List<JvmDeclaredType> nestedTypeDeclarators = session.getNestedTypeDeclarators();
		if (nestedTypeDeclarators.isEmpty())
			return Collections.emptyList();
		Iterator<JvmDeclaredType> iterator = nestedTypeDeclarators.iterator();
		String simpleName = name.getFirstSegment();
		List<IEObjectDescription> result = null;
		Set<JvmDeclaredType> nestedTypesSet = new HashSet<>();
		if (name.getSegmentCount() == 1) {
			while(iterator.hasNext()) {
				JvmDeclaredType declarator = iterator.next();
				Iterable<JvmDeclaredType> nestedTypes = declarator.findAllNestedTypesByName(simpleName);
				if (nestedTypes.iterator().hasNext()) {
					CollectionExtensions.addAll(nestedTypesSet, nestedTypes);
				}
			}
			if (nestedTypesSet.isEmpty()) {
				result = addDescriptions(name, 0, nestedTypesSet, result);
			}
		}
		if (result == null && name.getSegmentCount() == 1 && simpleName.indexOf('$') > 0) {
			QualifiedName splitted = QualifiedName.create(Strings.split(simpleName, '$'));
			return getLocalElementsByName(splitted);
		}
		if (result != null)
			return result;
		return Collections.emptyList();
	}
```

**Comment by szarnekow** [comment]
Very spooky code in NestedTypesScope ...
- `name.getSegmentCount() == 1` seems to guard all call to addDescriptions and can be moved to the top.

Do we have the same type twice in the list of nestedTypeDeclarators?

**Comment by szarnekow** [comment]
> name.getSegmentCount

Ahh you did this already. Sorry.

Yes, collecting the set of nested types - probably as LinkedHashSet to preserve order - should be fine.

**Comment by cdietrich** [comment]
no. in the sample above there is Sample and SampleImpl in  the list.
and cause sample impl inherits from Sample,
we report the nested types from Sample for SampleImpl too
findAllNestedTypesByName seems to collect parent types too

**Comment by cdietrich** [comment]
fixed in 2.19
